File: 卡码/代码随想录/1/11-图论/C++/graph/并查集/LeetCode_1971.cpp
//1971. 寻找图中是否存在路径
#include <iostream>
#include <vector>
using namespace std;

class Solution {
private:
	int n=200005;
	vector<int> father=vector<int>(n,0);
	void init(){
		for(int i=0;i<n;i++) father[i]=i;
	}
	int find(int u){
		return u==father[u] ? u : father[u]=find(father[u]);
	}
	bool isSame(int u,int v){
		u=find(u);
		v=find(v);
		return u==v;
	}
	void join(int u,int v){
		u=find(u);
		v=find(v);
		if(u==v) return;
		father[u]=v;
	}
public:
	bool validPath(int n, vector<vector<int>>& edges, int source, int destination) {
		init();
		for(int i=0;i<edges.size();i++){
			join(edges[i][0],edges[i][1]);
		}
		return isSame(source,destination);
	}
};

int main(){
  system("pause");
  return 0;
}

File: 卡码/代码随想录/1/11-图论/Python/并查集/LeetCode_1971.py
#1971. 寻找图中是否存在路径
from typing import *

class Solution:
  def __init__(self):
    self.father:List[int]=[]
  def init(self,n:int):
    self.father=[i for i in range(n)]
  def find(self,u:int)->int:
    if u==self.father[u]:
      return u
    else:
      self.father[u]=self.find(self.father[u])
      return self.father[u]
  def isSame(self,u:int,v:int):
    u=self.find(u)
    v=self.find(v)
    return u==v
  def join(self,u:int,v:int):
    u=self.find(u)
    v=self.find(v)
    if u==v:
      return
    self.father[u]=v
  def validPath(self, n: int, edges: List[List[int]], source: int, destination: int) -> bool:
    self.init(n)
    for i in range(len(edges)):
      self.join(edges[i][0],edges[i][1])
    return self.isSame(source,destination)